### ISSUE TITLE
Potential inconsistency when reading writing hdf strings

### ISSUE THREAD

**Issue by pgrete** [parthenon-hpc-lab/parthenon#515]
We recently encountered an interesting bug.

A regression test (run in the context of #511 ) consistently failed with 
```
4:     dataset[:,:,:,:,idx_component]
4: TypeError: 'NoneType' object is not subscriptable
```

We were able to isolate the issue that showed up on Python 3.8.x but not on Python 3.7.x after realizing that the default `h5py` version shipped in those environments made a major jump, e.g.,
```bash
$ python
Python 3.7.4 (default, Dec 22 2020, 12:15:41) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import h5py
>>> print(h5py.__file__)
/mnt/home/gretephi/.local/lib/python3.7/site-packages/h5py/__init__.py
>>> print(h5py.__version__)
2.10.0
```
versus
```bash
$ python
Python 3.8.2 (default, Feb 16 2021, 21:08:08) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import h5py
>>> print(h5py.__file__)
/mnt/home/gretephi/.local/lib/python3.8/site-packages/h5py/__init__.py
>>> print(h5py.__version__)
3.1.0
```

The bug showed up because the name comparison to identify a variable in the dataset failed as the names were read as byte array
```python
>>> datafile.Info["VariableNames"]
array([b'Density', b'MomentumDensity1', b'MomentumDensity2',
       b'MomentumDensity3', b'TotalEnergyDensity', b'Density',
       b'Velocity1', b'Velocity2', b'Velocity3', b'Pressure'],
      dtype=object)
```
which was a major change in `h5py` 3.x.x:
"_Variable-length UTF-8 strings in datasets are now read as bytes objects instead of str_", see https://docs.h5py.org/en/latest/whatsnew/3.0.html#breaking-changes-deprecations

We were able to fix the issue by  explicitly converting to string, e.g.,
```python
component_names = datafile.Info["VariableNames"].astype(str)
```

Given that the fix works I haven't looked into how we actually write the attributes and whether changing the encoding (when writing from Parthenon or reading with `phdf`) may potentially solve this automatically and/or make this more(?) consistent, see some additional details on string encoding here: https://docs.h5py.org/en/stable/strings.html#encodings

Bottom line: this issue may not even be  a real "issue" (given the easy fix above) and I just wanted to report it in some place in case other people stumble across it (which would then probably be the time to investigate further).

**Comment by jlippuner** [comment]
This is an annoying issue indeed. I'm not sure it's really a bug, though. From what I understand, it's a change in the `h5py` API that now spits out byte arrays instead of strings for string attributes, i.e. it's not a change in the HDF5 file itself that is being read and I don't think the HDF5 file could control how `h5py` chooses to represent the string attributes. As far as I know, there is only one type of string datatype in HDF5.

**Comment by Yurlungur** [comment]
Ah, I think I've encountered this in the past! And solved it the same way you did! My leaning is to just add the `astype` call into the hdf5 python reader and move on

**Comment by Yurlungur** [comment]
Resolved
